File: extensions/admob/template/android/libadmob/src/com/cocos/admob/service/RewardedAdService.java
package com.cocos.admob.service;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.cocos.admob.AdServiceHub;
import com.cocos.admob.core.Bridge;
import com.cocos.admob.core.IScriptHandler;
import com.cocos.admob.proto.rewarded.LoadRewardedAdACK;
import com.cocos.admob.proto.rewarded.LoadRewardedAdREQ;
import com.cocos.admob.proto.rewarded.OnUserEarnedRewardListenerNTF;
import com.cocos.admob.proto.rewarded.RewardedAdLoadCallbackNTF;
import com.cocos.admob.proto.rewarded.RewardedFullScreenContentCallbackNTF;
import com.cocos.admob.proto.rewarded.RewardedPaidEventNTF;
import com.cocos.admob.proto.rewarded.ShowRewardedAdACK;
import com.cocos.admob.proto.rewarded.ShowRewardedAdREQ;
import com.cocos.lib.CocosActivity;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdapterResponseInfo;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.OnUserEarnedRewardListener;
import com.google.android.gms.ads.rewarded.RewardItem;
import com.google.android.gms.ads.rewarded.RewardedAd;
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback;


/**
 * Internal class for rewarded ads.
 */
public final class RewardedAdService  extends  Service{
    private static final String TAG = "RewardedAdService";
    RewardedAd rewardedAd;
    String unitId;

    public void init(Bridge bridge, CocosActivity activity) {
        super.init(bridge, activity);

        bridge.getRoute().on(LoadRewardedAdREQ.class.getSimpleName(), LoadRewardedAdREQ.class, new IScriptHandler() {
            @Override
            public void onMessage(Object arg) {
                LoadRewardedAdREQ req = (LoadRewardedAdREQ) arg;
                loadAd(req.unitId);
                LoadRewardedAdACK ack = new LoadRewardedAdACK(unitId);
                bridge.sendToScript(LoadRewardedAdACK.class.getSimpleName(), ack);
            }
        });

        bridge.getRoute().on(ShowRewardedAdREQ.class.getSimpleName(), ShowRewardedAdREQ.class, new IScriptHandler() {
            @Override
            public void onMessage(Object arg) {
                ShowRewardedAdREQ req = (ShowRewardedAdREQ) arg;
                showAd();
                ShowRewardedAdACK ack = new ShowRewardedAdACK(unitId);
                bridge.sendToScript(ShowRewardedAdACK.class.getSimpleName(), ack);
            }
        });
    }

    private void loadAd(String unitId) {
        this.unitId = unitId;
        AdRequest adRequest = new AdRequest.Builder()
                .setRequestAgent(AdServiceHub.extensionVersion)
                .build();
        RewardedAd.load(activity, unitId,
                adRequest, new RewardedAdLoadCallback() {
                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                        // Handle the error.
                        Log.d(TAG, loadAdError.toString());
                        Toast.makeText(activity, "Rewarded ad failed to load.", Toast.LENGTH_SHORT).show();
                        rewardedAd = null;
                        bridge.sendToScript(RewardedAdLoadCallbackNTF.class.getSimpleName(), new RewardedAdLoadCallbackNTF(unitId, "onAdFailedToLoad", loadAdError.toString()));
                    }

                    @Override
                    public void onAdLoaded(@NonNull RewardedAd ad) {
                        rewardedAd = ad;
                        Log.d(TAG, "Ad was loaded.");
                        Toast.makeText(activity, "Rewarded ad loaded.", Toast.LENGTH_SHORT).show();
                        bridge.sendToScript(RewardedAdLoadCallbackNTF.class.getSimpleName(), new RewardedAdLoadCallbackNTF(unitId, "onAdLoaded"));

                        rewardedAd.setOnPaidEventListener(adValue -> {
                            RewardedPaidEventNTF rewardedPaidEventNTF = new RewardedPaidEventNTF(unitId);

                            rewardedPaidEventNTF.valueMicros = adValue.getValueMicros();
                            rewardedPaidEventNTF.currencyCode = adValue.getCurrencyCode();
                            rewardedPaidEventNTF.precision = adValue.getPrecisionType();

                            AdapterResponseInfo loadedAdapterResponseInfo = rewardedAd.getResponseInfo().
                                    getLoadedAdapterResponseInfo();
                            rewardedPaidEventNTF.adSourceName = loadedAdapterResponseInfo.getAdSourceName();
                            rewardedPaidEventNTF.adSourceId = loadedAdapterResponseInfo.getAdSourceId();
                            rewardedPaidEventNTF.adSourceInstanceName = loadedAdapterResponseInfo.getAdSourceInstanceName();
                            rewardedPaidEventNTF.adSourceInstanceId = loadedAdapterResponseInfo.getAdSourceInstanceId();

                            Bundle extras = rewardedAd.getResponseInfo().getResponseExtras();
                            rewardedPaidEventNTF.mediationGroupName = extras.getString("mediation_group_name");
                            rewardedPaidEventNTF.mediationABTestName = extras.getString("mediation_ab_test_name");
                            rewardedPaidEventNTF.mediationABTestVariant = extras.getString("mediation_ab_test_variant");

                            bridge.sendToScript(RewardedPaidEventNTF.class.getSimpleName(), rewardedPaidEventNTF);
                        });

                        rewardedAd.setFullScreenContentCallback(new FullScreenContentCallback() {
                            @Override
                            public void onAdClicked() {
                                // Called when a click is recorded for an ad.
                                Log.d(TAG, "Ad was clicked.");
                                bridge.sendToScript(RewardedFullScreenContentCallbackNTF.class.getSimpleName(), new RewardedFullScreenContentCallbackNTF(unitId, "onAdClicked"));
                            }

                            @Override
                            public void onAdDismissedFullScreenContent() {
                                // Called when ad is dismissed.
                                // Set the ad reference to null so you don't show the ad a second time.
                                Log.d(TAG, "Ad dismissed fullscreen content.");
                                rewardedAd = null;
                                bridge.sendToScript(RewardedFullScreenContentCallbackNTF.class.getSimpleName(), new RewardedFullScreenContentCallbackNTF(unitId, "onAdDismissedFullScreenContent"));
                            }

                            @Override
                            public void onAdFailedToShowFullScreenContent(AdError adError) {
                                // Called when ad fails to show.
                                Log.e(TAG, "Ad failed to show fullscreen content.");
                                rewardedAd = null;
                                bridge.sendToScript(RewardedFullScreenContentCallbackNTF.class.getSimpleName(), new RewardedFullScreenContentCallbackNTF(unitId, "onAdFailedToShowFullScreenContent", adError.toString()));
                            }

                            @Override
                            public void onAdImpression() {
                                // Called when an impression is recorded for an ad.
                                Log.d(TAG, "Ad recorded an impression.");
                                bridge.sendToScript(RewardedFullScreenContentCallbackNTF.class.getSimpleName(), new RewardedFullScreenContentCallbackNTF(unitId, "onAdImpression"));
                            }

                            @Override
                            public void onAdShowedFullScreenContent() {
                                // Called when ad is shown.
                                Log.d(TAG, "Ad showed fullscreen content.");
                                bridge.sendToScript(RewardedFullScreenContentCallbackNTF.class.getSimpleName(), new RewardedFullScreenContentCallbackNTF(unitId, "onAdShowedFullScreenContent"));
                            }
                        });
                    }
                });
    }

    private void showAd() {
        Log.d(TAG, "showAd: ");
        if (rewardedAd == null) {
            Log.d(TAG, "The rewarded ad wasn't ready yet.");
            return;
        }

        Activity activityContext = activity;
        rewardedAd.show(activityContext, new OnUserEarnedRewardListener() {
            @Override
            public void onUserEarnedReward(@NonNull RewardItem rewardItem) {
                // Handle the reward.
                Log.d(TAG, "The user earned the reward.");
                int rewardAmount = rewardItem.getAmount();
                String rewardType = rewardItem.getType();
                bridge.sendToScript(OnUserEarnedRewardListenerNTF.class.getSimpleName(), new OnUserEarnedRewardListenerNTF(unitId, rewardType, rewardAmount));
            }
        });
    }
}

File: extensions/GoogleAdMob/template/ios/admob/service/RewardedAdService.m
/****************************************************************************
 Copyright (c) 2023-2024 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos.com
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You
 shall not use Cocos Creator software for developing other software or tools
 that's used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.
 
 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to
 you.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#import "RewardedAdService.h"

#import <GoogleMobileAds/GoogleMobileAds.h>

#import "Route.h"

#import "LoadRewardedAdACK.h"
#import "LoadRewardedAdREQ.h"
#import "OnUserEarnedRewardListenerNTF.h"
#import "RewardedAdLoadCallbackNTF.h"
#import "RewardedFullScreenContentCallbackNTF.h"
#import "RewardedPaidEventNTF.h"
#import "ShowRewardedAdACK.h"
#import "ShowRewardedAdREQ.h"

#import "../AdServiceHub.h"

@interface RewardedAdService()<GADFullScreenContentDelegate>

@property(nonatomic, strong) GADRewardedAd *rewardedAd;
@property(nonatomic, strong) NSString *unitId;

@property (nonatomic, weak) Bridge *bridge;

@end

@implementation RewardedAdService

- (instancetype)initWithBridge:(Bridge *)bridge {
    self = [super init];
    if (self) {
        self.bridge = bridge;
        
        __weak typeof(self) wself = self;
        // load
        [bridge.route on:[LoadRewardedAdREQ class].description type:[LoadRewardedAdREQ class] messageHandler:^(id arg) {
            LoadRewardedAdREQ *req = (LoadRewardedAdREQ *)arg;
            [wself loadAd:req.unitId];
            LoadRewardedAdACK *ack = [[LoadRewardedAdACK alloc] initWithUnitId:req.unitId];
            [bridge sendToScript:[LoadRewardedAdACK class].description src:ack];
        }];
        
        // show
        [bridge.route on:[ShowRewardedAdREQ class].description type:[ShowRewardedAdREQ class] messageHandler:^(id arg) {
            ShowRewardedAdREQ *req = (ShowRewardedAdREQ *)arg;
            [wself showAd];
            ShowRewardedAdACK *ack = [[ShowRewardedAdACK alloc] initWithUnitId:req.unitId];
            [bridge sendToScript:[ShowRewardedAdACK class].description src:ack];
        }];
    }
    return self;
}

- (void)loadAd:(NSString *)unitId {
    self.unitId = unitId;
    GADRequest *request = [GADRequest request];
    request.requestAgent = [[AdServiceHub sharedInstance] extensionVersion];
    [GADRewardedAd
     loadWithAdUnitID:unitId
     request:request
     completionHandler:^(GADRewardedAd *ad, NSError *error) {
        if (error) {
            NSLog(@"Rewarded ad failed to load with error: %@", [error localizedDescription]);
            self.rewardedAd = nil;
            
            RewardedAdLoadCallbackNTF *ntf = [[RewardedAdLoadCallbackNTF alloc] initWithUnitId:unitId method:@"onAdFailedToLoad" loadAdError:[error localizedDescription]];
            [self.bridge sendToScript:[RewardedAdLoadCallbackNTF class].description src:ntf];
            return;
        }
        self.rewardedAd = ad;
        
        RewardedAdLoadCallbackNTF *ntf = [[RewardedAdLoadCallbackNTF alloc] initWithUnitId:unitId method:@"onAdLoaded"];
        [self.bridge sendToScript:[RewardedAdLoadCallbackNTF class].description src:ntf];
        
        self.rewardedAd.fullScreenContentDelegate = self;
        
        //paidEventHandler
        __weak typeof(self) wself = self;
        __weak GADRewardedAd *weakRewareded = ad;
        self.rewardedAd.paidEventHandler = ^void(GADAdValue *_Nonnull adValue) {
            RewardedPaidEventNTF *ntf = [[RewardedPaidEventNTF alloc] initWithUnitId:weakRewareded.adUnitID];
            
            ntf.valueMicros = [[adValue value] longValue];
            ntf.currencyCode = adValue.currencyCode;
            ntf.precision = (int)adValue.precision;
            
            GADAdNetworkResponseInfo *loadedAdNetworkResponseInfo = weakRewareded.responseInfo.loadedAdNetworkResponseInfo;
            ntf.adSourceName = loadedAdNetworkResponseInfo.adSourceName;
            ntf.adSourceId = loadedAdNetworkResponseInfo.adSourceID;
            ntf.adSourceInstanceName = loadedAdNetworkResponseInfo.adSourceInstanceName;
            ntf.adSourceInstanceId = loadedAdNetworkResponseInfo.adSourceInstanceID;
            
            NSDictionary<NSString *, id> *extras = weakRewareded.responseInfo.extrasDictionary;
            ntf.mediationGroupName = extras[@"mediation_group_name"];
            ntf.mediationABTestName = extras[@"mediation_ab_test_name"];
            ntf.mediationABTestVariant = extras[@"mediation_ab_test_variant"];
            
            [wself.bridge sendToScript:[RewardedPaidEventNTF class].description src:ntf];
        };
    }];
}

- (void)showAd {
    if(!self.rewardedAd) {
        NSLog(@"rewarded ad is not loaded");
        return;
    }
    NSSet<UIScene *> *connectedScenes = UIApplication.sharedApplication.connectedScenes;
    NSMutableArray<UIWindowScene *> *windowScenes = [NSMutableArray array];
    for (UIScene *scene in connectedScenes) {
        if ([scene isKindOfClass:[UIWindowScene class]]) {
            [windowScenes addObject:(UIWindowScene *)scene];
        }
    }
    UIViewController *viewController = nil;
    for (UIWindowScene *scene in windowScenes) {
        if (scene.activationState == UISceneActivationStateForegroundActive) {
            UIWindow *keyWindow = scene.windows.firstObject;
            UIViewController *rootViewController = keyWindow.rootViewController;
            if (!rootViewController) {
                return;
            }
            viewController = rootViewController;
            break;
        }
    }
    if(!viewController) {
        NSLog(@"interstitial service viewController search failure!");
        return;
    }
    
    [self.rewardedAd presentFromRootViewController:viewController
                          userDidEarnRewardHandler:^{
        GADAdReward *reward = self.rewardedAd.adReward;
        OnUserEarnedRewardListenerNTF *ntf = [[OnUserEarnedRewardListenerNTF alloc]
                                                         initWithUnitId:self.unitId
                                                         rewardType:reward.type
                                                         rewardamount:[reward.amount intValue]];
        [self.bridge sendToScript:[OnUserEarnedRewardListenerNTF class].description src:ntf];
    }];
}

#pragma mark GADFullScreenContentDelegate implementation

/// Tells the delegate that an impression has been recorded for the ad.
- (void)adDidRecordImpression:(nonnull id<GADFullScreenPresentingAd>)ad {
    RewardedFullScreenContentCallbackNTF *ntf = [[RewardedFullScreenContentCallbackNTF alloc]
                                                     initWithUnitId:self.unitId
                                                     method:@"onAdImpression"];
    [self.bridge sendToScript:[RewardedFullScreenContentCallbackNTF class].description src:ntf];
}

/// Tells the delegate that a click has been recorded for the ad.
- (void)adDidRecordClick:(nonnull id<GADFullScreenPresentingAd>)ad {
    RewardedFullScreenContentCallbackNTF *ntf = [[RewardedFullScreenContentCallbackNTF alloc]
                                                     initWithUnitId:self.unitId
                                                     method:@"onAdClicked"];
    [self.bridge sendToScript:[RewardedFullScreenContentCallbackNTF class].description src:ntf];
}

/// Tells the delegate that the ad failed to present full screen content.
- (void)ad:(nonnull id<GADFullScreenPresentingAd>)ad
didFailToPresentFullScreenContentWithError:(nonnull NSError *)error {
    self.rewardedAd = nil;
    RewardedFullScreenContentCallbackNTF *ntf = [[RewardedFullScreenContentCallbackNTF alloc]
                                                     initWithUnitId:self.unitId
                                                     method:@"onAdFailedToShowFullScreenContent"
                                                     adError:[error localizedDescription]];
    [self.bridge sendToScript:[RewardedFullScreenContentCallbackNTF class].description src:ntf];
}

/// Tells the delegate that the ad will present full screen content.
- (void)adWillPresentFullScreenContent:(nonnull id<GADFullScreenPresentingAd>)ad {
    RewardedFullScreenContentCallbackNTF *ntf = [[RewardedFullScreenContentCallbackNTF alloc]
                                                     initWithUnitId:self.unitId
                                                     method:@"onAdShowedFullScreenContent"];
    [self.bridge sendToScript:[RewardedFullScreenContentCallbackNTF class].description src:ntf];
}

/// Tells the delegate that the ad will dismiss full screen content.
- (void)adWillDismissFullScreenContent:(nonnull id<GADFullScreenPresentingAd>)ad {
    
}

/// Tells the delegate that the ad dismissed full screen content.
- (void)adDidDismissFullScreenContent:(nonnull id<GADFullScreenPresentingAd>)ad {
    self.rewardedAd = nil;
    RewardedFullScreenContentCallbackNTF *ntf = [[RewardedFullScreenContentCallbackNTF alloc]
                                          initWithUnitId:self.unitId
                                          method:@"onAdDismissedFullScreenContent"];
    [self.bridge sendToScript:[RewardedFullScreenContentCallbackNTF class].description src:ntf];
}

@end


